base/multidimensional.jl

```diff
@@ -178,100 +172,164 @@ end  # IteratorsMD
 
 using .IteratorsMD
 
+### From abstractarray.jl: Internal multidimensional indexing definitions ###
+# These are not defined on directly ongetindex and unsafe_getindex to avoid
+# ambiguities for AbstractArray subtypes. See the note in abstractarray.jl
 
-### From array.jl
-
-stagedfunction checksize(A::AbstractArray, I...)
+# Note that it's most efficient to call checkbounds first, and then to_index
+stagedfunction _getindex(l::LinearIndexing, A::AbstractArray, I::Union(Real, AbstractArray, Colon)...)
+    Isplat = Expr[:(I[$d]) for d = 1:length(I)]
+    quote
+        $(Expr(:meta, :inline))
+        checkbounds(A, $(Isplat...))
+        _unsafe_getindex(l, A, $(Isplat...))
+    end
+end
+stagedfunction _unsafe_getindex(l::LinearIndexing, A::AbstractArray, I::Union(Real, AbstractArray, Colon)...)
     N = length(I)
     quote
-        @nexprs $N d->(size(A, d) == length(I[d]) || throw(DimensionMismatch("index $d has length $(length(I[d])), but size(A, $d) = $(size(A,d))")))
-        nothing
+        $(Expr(:meta, :inline))
+        @nexprs $N d->(I_d = to_index(I[d]))
+        dest = similar(A, @ncall $N index_shape A I)
+        @ncall $N checksize dest I
+        @ncall $N _unsafe_getindex! linearindexing(dest) dest l A I
+    end
+end
+
+# logical indexing optimization - don't use find (within to_index)
+# Todo: use magic to speed up LinearSlow src
+@inline function _unsafe_getindex(::LinearIndexing, src::AbstractArray, I::AbstractArray{Bool})
+    # Both index_shape and checksize compute sum(I); manually hoist it out
+    N = sum(I)
+    dest = similar(src, (N,))
+    size(dest) == (N,) || throw(DimensionMismatch())
```

Comment:
Not really - `similar` has.  It could give us the wrong answer, and the check here is very cheap compared to the allocations and many assignments (although I may need to hide it in a function; I haven't profiled yet).  In general, my approach has been to not trust the output of `similar`. Master effectively does the same thing for arrays (although it's broken across method boundaries): https://github.com/JuliaLang/julia/blob/master/base/multidimensional.jl#L219